source/common/upstream/upstream_impl.cc

```diff
@@ -899,8 +899,16 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(
     }
 
     auto existing_host = all_hosts_.find(host->address()->asString());
+    const bool existing_host_found = existing_host != all_hosts_.end();
 
-    if (existing_host != all_hosts_.end()) {
+    // If we have found a new host matched, based on address, to an existing one: check if the
+    // health check address of that host is different. If it is, we need to rebuild. Note that this
+    // checking matters only if the cluster's active health checker is activated.
+    const bool health_check_changed =
+        health_checker_ != nullptr && existing_host_found &&
+        *existing_host->second->healthCheckAddress() != *host->healthCheckAddress();
+
+    if (existing_host_found && !health_check_changed) {
```

Comment:
I think this can still be simplified a bit, since later on, we might add another rebuild condition. I'd rename `health_check_changed` to `no_inplace_host_update`. Also, maybe add a comment to the top of this `if` block explaining what it's doing in terms of hosts insertion/deletion behavior.